nilearn/input_data/nifti_spheres_masker.py

```diff
@@ -85,6 +86,94 @@ def _apply_mask_and_get_affinity(seeds, niimg, radius, allow_overlap,
+    target_shape space.
+
+    Parameters
+    ----------
+    seeds_: List of triplets of coordinates in native space
+        Seed definitions. List of coordinates of the seeds in the same space
+        as target_affine.
+    radius: float
+        Indicates, in millimeters, the radius for the sphere around the seed.
+    allow_overlap: boolean
+        If False, an error is raised if the maps overlaps (ie at least two
+        maps have a non-zero value for the same voxel).
+    mask_img: Niimg-like object,
+        Mask as a reference space to project the spheres back into brain space.
+
+    Returns
+    -------
+    spheres_map: list of lists 
+        The spheres' location (flattened array), one list for each sphere.
+    '''
+    seeds = list(seeds)
+
+    # Load masking data
+    mask, affine = masking._load_mask_img(mask_img)
+
+    # Get coordinate for alle voxels inside of mask
+    brain_coords = list(np.ndindex(*mask.shape))
+
+    # Get the closest voxels for each seed
+    nearests = []
+    for sx, sy, sz in seeds:
+        nearest = np.round(coord_transform(sx, sy, sz, 
+                           np.linalg.inv(affine)))
+        nearest = nearest.astype(int)
+        nearest = (nearest[0], nearest[1], nearest[2])
+        try:
+            nearests.append(brain_coords.index(nearest))
+        except ValueError:
+            nearests.append(None)
+
+    brain_coords = np.asarray(list(zip(*brain_coords)))
+    # Calculate mask and the corresponding index in flattended array.
+    mask_coords = list(zip(*np.where(mask == 1)))
+    mask_coords = np.asarray(mask_coords).astype(int).T
+    mask_idx = np.vstack([np.in1d(brain_coords[0, :], mask_coords[0, :]),
+                          np.in1d(brain_coords[1, :], mask_coords[1, :]),
+                          np.in1d(brain_coords[2, :], mask_coords[2, :])]
+                        ).all(0)
+
+    # Bring brain_coords into the masks space.
+    brain_coords = coord_transform(brain_coords[0], brain_coords[1],
+                                  brain_coords[2], affine)
+    brain_coords = np.asarray(brain_coords).T
+    # Calculate the adjacent coordinates in radius.
+    clf = neighbors.NearestNeighbors(radius=radius)
+    adjacency = clf.fit(brain_coords).radius_neighbors_graph(seeds)
+    adjacency = adjacency.tolil()
+
+    # Include the voxel containing the seed itself
+    brain_coords = brain_coords.astype(int).tolist()
+
+    for i, (seed, nearest) in enumerate(zip(seeds, nearests)):
+        # Include closest voxel
+        if nearest is not None:
+            adjacency[i, nearest] = True
+        try:
+            # Include seed itself
+            adjacency[i, brain_coords.index(np.array(seed))] = True
+        except ValueError:
+            # seed is not in the mask
+            pass
+
+    # Removing voxels not inside of the mask.
+    adjacency[:, mask_idx==False] = False
+    
+    if np.any(adjacency.sum(axis=0) >= 2):
```

Comment:
```suggestion
    adjacency = adjacency.tocsr()
    if np.any(adjacency.sum(axis=0) >= 2):
```
LIL format is good for constructing the matrix but before performing linear operations on it we must convert it to a more efficient format such as CSR, see [the scipy documentation](https://docs.scipy.org/doc/scipy/reference/sparse.html#usage-information)

Reply:
Applying `.tocsr()` here breaks the function. I unfortunately have no idea about CSR format etc. but it doesn't seem to have `.rows` Attribute which is later used in the iteration. 

I am not sure if transforming back and forth will help much.

Reply:
we can use `.getrow(i)` and iterate over row indices instead of using the LIL's `.rows` and it will be faster. otherwise, yes transforming back and forth would help. basically the lil (list of lists) format contains for each row, the list of non-zero column indices, and the list of corresponding values. adding rows or changing the sparsity structure is cheap but arithmetic operations are expensive. So we convert to CSR (compressed sparse row), which contains one array of column indices, one array of indices into the first array indicating where each row begins, and one array of values. adding one non-zero can change everything in a CSR matrix so it is not efficient for constructing a sparse matrix by adding elements as we do here, but it is very efficient for computations. this is why a very common scheme is to use a LIL to construct the matrix, then convert to CSR or CSC